ffi/tests/firewood/merkle_compatibility_test.go

```diff
@@ -0,0 +1,442 @@
+	pr := tr.selectRandomProposal()
+
+	if pr == nil {
+		return
+	}
+
+	batchSize := tr.rand.Intn(maxBatchSize) + 1 // ensure at least one key-value pair
+	keys, vals := tr.createRandomBatch(batchSize)
+
+	fwdPr := pr.fwdView
+	merkleView := pr.merkleView
+
+	fwdChildPr, err := fwdPr.Propose(keys, vals)
+	tr.require.NoError(err)
+
+	merkleViewChange := merkledb.ViewChanges{}
+	for i := range keys {
+		merkleViewChange.BatchOps = append(merkleViewChange.BatchOps, database.BatchOp{
+			Key:   keys[i],
+			Value: vals[i],
+		})
+	}
+	merkleChildView, err := merkleView.NewView(context.Background(), merkleViewChange)
+	tr.require.NoError(err)
+
+	fwdRoot, err := fwdChildPr.Root()
+	tr.require.NoError(err)
+	merkleRoot, err := merkleChildView.GetMerkleRoot(context.Background())
+	tr.require.NoError(err)
+	tr.require.Equal(fwdRoot, merkleRoot[:])
+
+	tr.nextID++
+	newProposal := &proposal{
+		parentID:   pr.id,
+		id:         tr.nextID,
+		merkleView: merkleChildView,
+		fwdView:    fwdChildPr,
+	}
+	pr.children = append(pr.children, newProposal)
+	tr.proposals[newProposal.id] = newProposal
+}
+
+func (tr *tree) createProposalOnDB() {
+	batchSize := tr.rand.Intn(maxBatchSize) + 1 // ensure at least one key-value pair
+	keys, vals := tr.createRandomBatch(batchSize)
+	fwdPr, err := tr.fwdDB.Propose(keys, vals)
+	tr.require.NoError(err)
+
+	merkleViewChange := merkledb.ViewChanges{}
+	for i := range keys {
+		merkleViewChange.BatchOps = append(merkleViewChange.BatchOps, database.BatchOp{
+			Key:   keys[i],
+			Value: vals[i],
+		})
+	}
+	merkleChildView, err := tr.merkleDB.NewView(context.Background(), merkleViewChange)
+	tr.require.NoError(err)
+
+	fwdRoot, err := fwdPr.Root()
+	tr.require.NoError(err)
+	merkleRoot, err := merkleChildView.GetMerkleRoot(context.Background())
+	tr.require.NoError(err)
+	tr.require.Equal(fwdRoot, merkleRoot[:])
+
+	tr.nextID++
+	newProposal := &proposal{
+		parentID:   tr.id,
+		id:         tr.nextID,
+		merkleView: merkleChildView,
+		fwdView:    fwdPr,
+	}
+	tr.children = append(tr.children, newProposal)
+	tr.proposals[newProposal.id] = newProposal
+}
+
+func (tr *tree) proposalGet() {
+	pr := tr.selectRandomProposal()
+	if pr == nil {
+		return
+	}
+
+	keys := tr.selectRandomKeys(maxBatchSize)
+	for _, key := range keys {
+		fwdVal, fwdErr := pr.fwdView.Get(key)
+		merkleVal, merkleErr := pr.merkleView.GetValue(context.Background(), key)
+
+		// XXX: handling of non-existent keys diverges between merkleDB and firewood
+		if merkleErr == database.ErrNotFound {
+			tr.require.Nil(merkleVal)
+			tr.require.NoError(fwdErr)
+			tr.require.Nil(fwdVal)
+			continue
+		}
+		tr.require.NoError(merkleErr)
+		tr.require.NoError(fwdErr)
+		tr.require.Equal(fwdVal, merkleVal)
+	}
+}
+
+func (tr *tree) commitProposal() {
+	if len(tr.children) == 0 {
+		return // no proposals to commit
+	}
+	tr.commitRandomProposal()
+}
+
+func fuzzTree(t *testing.T, randSource int64, byteSteps []byte) {
+	rand := rand.New(rand.NewSource(randSource))
```

Comment:
Do we have some way of "remembering" this random value, so we can recreate the same run again if it fails?